fractional-NFT/project/token-distributor-contract/src/interface.sw

```diff
@@ -0,0 +1,162 @@
+library interface;
+
+dep data_structures;
+
+use data_structures::TokenDistribution;
+
+abi TokenDistributor {
+    /// Allows for the token admin to start a buyback of tokens at a set price.
+    ///
+    /// # Arguments
+    ///
+    /// * `fractional_nft` - The token contract that holds the NFT.
+    /// * `token_price` - The price for a single fractionalized NFT token.
+    ///
+    /// # Reverts
+    ///
+    /// * When `fractional_nft` does not map to an exising token distribution.
+    /// * When the sender is not the token admin.
+    /// * When the distribution is not in the distributing state.
+    /// * When providing an incorrect payment asset or amount.
+    #[storage(read, write)]
+    fn buyback(fractional_nft: ContractId, token_price: u64);
+
+    /// Starts a new token distribution and takes control of the NFT.
+    ///
+    /// # Arguments
+    ///
+    /// * `external_asset` - The asset which will be accepted in return for fractionalized NFT tokens.
+    /// * `fractional_nft` - The token contract that holds the NFT.
+    /// * `nft` - The contract that manages the NFT held by the fractionalized NFT.
+    /// * `reserve_price` - The price at which admin rights of the NFT may be bought outright.
+    /// * `token_admin` - The user which will have the ability to withdraw both the paid tokens and the NFT.
+    /// * `token_price` - The price for a single fractionalized NFT token.
+    /// * `token_supply` - The total number of fractionalized NFT tokens which will be minted.
+    /// * `nft_token_id` - The id of the NFT which will be fractionalized.
+    ///
+    /// # Reverts
+    ///
+    /// * When `fractional_nft` maps to an already existing token distribution.
+    #[storage(read, write)]
+    fn create(external_asset: ContractId, fractional_nft: ContractId, nft: ContractId, reserve_price: Option<u64>, token_admin: Option<Identity>, token_price: u64, token_supply: u64, nft_token_id: u64);
+
+    /// Will end the token distribution and return the NFT.
+    ///
+    /// # Arguments
+    ///
+    /// * `fractional_nft` - The token contract that holds the NFT.
+    ///
+    /// # Reverts
+    ///
+    /// * When the `fractional_nft` does not map to an existing token distribution.
+    /// * When the sender is not the admin of the distribution.
+    /// * When not all tokens can be returned to the fractionalized NFT contract.
```

Comment:
I think it would be better to also mention the state because otherwise there in an implied step that must be taken to know when this will revert